### DISCUSSION TITLE
Pyright and Mypy divergence on partially overlapping overloads

### DISCUSSION THREAD

**Discussion by insilications** [microsoft/pyright#6824] [General]
All examples tested with Python 3.12. Both Mypy and Pyright can return an error when there are (1) overloaded functions overlapping and (2) they return incompatible return types. I have found some differences in this situation. My conclusion is that Pyright is doing the right thing here, but I want confirmation so I can open an issue on Mypy: 
    

```python
from __future__ import annotations
from typing import overload

class BaseClass:
    @overload
    def select(self, arg1: tuple[BaseClass | str | tuple[bool, bool], BaseClass | str], /) -> str: ...
    @overload
    # def select(self, arg1: tuple[BaseClass | str | tuple[bool, bool] | int | float, BaseClass | str | int | float], /) -> BaseClass: ...
    def select(self, arg1: tuple[tuple[bool, bool] | int | float, BaseClass | str | int | float], /) -> BaseClass: ... # remove the first overlapping "BaseClass | str"
```

In the latest upstream Pyright, this solves the `Overload 1 overlaps overload 2 and returns an incompatible type` error.
But on latest upstream Mypy, the error persists: `Overloaded function signatures 1 and 2 overlap with incompatible return types`.
(https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=8851e47459487d604257ab318a9457ce)

```python
from __future__ import annotations
from typing import overload

class BaseClass:
    @overload
    def select(self, arg1: tuple[BaseClass | str | tuple[bool, bool], BaseClass | str], /) -> str: ...
    @overload
    # def select(self, arg1: tuple[BaseClass | str | tuple[bool, bool] | int | float, BaseClass | str | int | float], /) -> BaseClass: ...
    # def select(self, arg1: tuple[tuple[bool, bool] | int | float, BaseClass | str | int | float], /) -> BaseClass: ... # remove the first overlapping "BaseClass | str"
    def select(self, arg1: tuple[tuple[bool, bool] | int | float, int | float], /) -> BaseClass: ... # remove the first and the second overlapping "BaseClass | str"
```

To fully solve both overloads on latest upstream Mypy, you need to remove the first and the second overlapping "BaseClass | str".

PEP 484 isn't clear on how overloads should work, but I think Pyright is correct. The Parameter 2 of the overloads have partially overlapping Union types. For Pyright, removing just the first partial overlap in the first Union is enough to disambiguate between the two overloads. And that is the correct inference, I think.

**Comment by Viicos** [comment]
overload resolution being under specified, there is some discussion here: https://discuss.python.org/t/what-is-the-future-of-overload-resolution/41666/2.

I think an issue @ mypy could be opened if you feel pyright is doing the right thing, but it might be worth updating the typing spec first.

As a start, I would like to collect the overload resolution behaviors from (at least) mypy and pyright, find where it is diverging, and create a list of points to be discussed from these diverging cases. As said in the Python discussion, mypy is sometimes inconsistent with its own documentation, I would also like to find such cases so that they can also be discussed.

I'm a bit busy however and don't have much knowledge regarding the internals of mypy/pyright, so it might take some time.

**Reply by erictraut** [reply]
I think mypy's error is correct here. These overloads are overlapping with incompatible return types. What overload would you expect if you passed an argument of type `tuple[tuple[bool, bool], BaseClass | str]` for `arg1`? Both overloads would match in this case, and they return different types.

Pyright's logic for detecting overlapping overloads should be smarter about unions.

Here's a simplified example where the divergence can be seen.

```python
class BaseClass:
    @overload
    def select(self, arg1: int | set[int], /) -> str:
        ...

    @overload
    def select(self, arg1: int | list[int], /) -> BaseClass:
        ...
```

I've created a [bug report](https://github.com/microsoft/pyright/issues/6825) to track this issue.

**Reply by insilications** [reply]
I was sleepy. You are correct... Removing just the first partial overlap in the first Union is NOT enough to disambiguate between the two overloads. 

I've opened https://github.com/microsoft/pyright/pull/6846 to try to fix the more specific case regarding tuples with overlapping unions. I don't know if modifying `assignTupleTypeArguments` in such a way is going to create side effects. Fixing https://github.com/microsoft/pyright/issues/6825 is going to require more complex changes AFAICT.
